ntp-proto/src/nts_record.rs

```diff
@@ -366,3 +379,418 @@ impl NtsRecordDecoder {
         }
     }
 }
+
+#[derive(Debug)]
+pub enum KeyExchangeError {
+    UnrecognizedCriticalRecord,
+    BadRequest,
+    InternalServerError,
+    UnknownErrorCode(u16),
+    NoValidProtocol,
+    NoValidAlgorithm,
+    NoCookies,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct KeyExchange {
+    pub remote: String,
+    pub port: u16,
+    pub cookies: Vec<Vec<u8>>,
+}
+
+pub type KeyExchangeStep = ControlFlow<Result<KeyExchange, KeyExchangeError>, KeyExchange>;
+
+pub enum NextStep {
+    NeedMoreInput(KeyExchange),
+    Done(KeyExchange),
+    ReadError(std::io::Error),
+    KeyExchangeError(KeyExchangeError),
+}
```

Comment:
Looking at this whole enum, I think it would be better to make this a result type. Also, for the parser, we would provide a nicer interface I think if the final result (after Done) is the only place where data can be extracted from, as that would prevent accidentally using a partial exchange result.